src/ClientData/CallstackDataTest.cpp

```diff
@@ -161,43 +170,178 @@ TEST(CallstackData, FilterCallstackEventsBasedOnMajorityStartExcludesFunctionToS
+                    [&](int id) { return list[id]; });
+  return subset;
+}
+
+struct FormattingForEachCallstackEventOfTidInTimeRangeDiscretizedTest {
+  std::string test_name;
+  uint32_t tid;
+  uint64_t start_ns;
+  uint64_t end_ns;
+  uint32_t resolution;
+  std::vector<int> expected_event_ids;
+};
+
+using ForEachCallstackEventOfTidInTimeRangeDiscretizedTest =
+    TestWithParam<FormattingForEachCallstackEventOfTidInTimeRangeDiscretizedTest>;
+
+TEST_P(ForEachCallstackEventOfTidInTimeRangeDiscretizedTest, IterationIsCorrect) {
+  const FormattingForEachCallstackEventOfTidInTimeRangeDiscretizedTest& test_case = GetParam();
+
+  CallstackData callstack_data;
+  const CallstackInfo cs{{0x11, 0x10}, CallstackType::kComplete};
+  callstack_data.AddUniqueCallstack(kCallstackId1, std::move(cs));
+
+  for (const CallstackEvent& event : kAllEvents) {
+    callstack_data.AddCallstackEvent(event);
+  }
+
+  std::vector<CallstackEvent> visited_callstack_list;
+  auto visit_callstack = [&](const CallstackEvent& event) {
+    visited_callstack_list.push_back(event);
+  };
+  callstack_data.ForEachCallstackEventOfTidInTimeRangeDiscretized(
+      test_case.tid, test_case.start_ns, test_case.end_ns, test_case.resolution, visit_callstack);
   EXPECT_THAT(
-      callstack_data.GetCallstackEventsOfTidInTimeRange(kTid, 0,
-                                                        std::numeric_limits<uint64_t>::max()),
-      testing::Pointwise(CallstackEventEq(), std::vector<CallstackEvent>{event1, event2, event3,
-                                                                         event4, event5, event6}));
+      visited_callstack_list,
+      Pointwise(CallstackEventEq(), GetSubset(kAllEvents, test_case.expected_event_ids)));
 }
 
+constexpr uint64_t kStartNs = 0;
+constexpr uint64_t kEndNs = 1000;
+constexpr uint64_t kMaxNs = std::numeric_limits<uint64_t>::max();
+constexpr uint32_t kResolution = 2000;
+
+INSTANTIATE_TEST_SUITE_P(
+    ForEachCallstackEventOfTidInTimeRangeDiscretizedTests,
+    ForEachCallstackEventOfTidInTimeRangeDiscretizedTest,
+    ValuesIn<FormattingForEachCallstackEventOfTidInTimeRangeDiscretizedTest>({
+        {"NormalTimeRange", kTid, kStartNs, kEndNs, kResolution, {0, 1, 3}},
+        {"DifferentTid", kAnotherTid, kStartNs, kEndNs, kResolution, {2}},
+        {"SmallTimeRange", kTid, kStartNs, kTimestamps[2] - 1, kResolution, {0, 1}},
+        // When max_timestamp is std::numeric_limits<uint64_t>::max(), each callstack will be drawn
+        // in the first pixel, and therefore only one will be visible.
+        {"InfiniteTimeRange", kTid, kStartNs, kMaxNs, kResolution, {0}},
+        // With one pixel on the screen we should only see one event.
+        {"OnePixel", kTid, kStartNs, kEndNs, /*resolution=*/1, {0}},
+    }),
+    [](const TestParamInfo<ForEachCallstackEventOfTidInTimeRangeDiscretizedTest::ParamType>& info) {
+      return info.param.test_name;
+    });
+
+enum ExpectType { kExpectAll, kExpectAny };
```

Comment:
Just a thought. 

I don't like this enum. One way to circumvent the need for it would be to introduce two lambdas

```
constexpr auto kExpectAll = [](const auto& actual, const auto& expected) {
  EXPECT_THAT(actual, Pointwise(CallstackEventEq(), expected));
};

constexpr auto kExpectAny = [](const auto& actual, const auto& expected) {
  EXPECT_EQ(actual.size(), 1);
  EXPECT_THAT(actual[0], AnyOfArray(expected));
};
```
And replace the field `ExpectType expect_type` in `FormattingForEachCallstackEventInTimeRangeDiscretizedTest` with `absl::FunctionRef<void(std::vector<CallstackEvent>, std::vector<CallstackEvent>)> expect;`. Then in the test you can just call 
```
test_case.expect(visited_callstack_list, GetSubset(kAllEvents, test_case.expected_event_ids));
```